File: NLP_HW1/js/fsa.js
function fsa(transition , start_state , fin_state , speech_list){
  this.transition = transition;
  this.start_state = start_state;
  this.fin_state  = fin_state;
  this.speech_list = speech_list;
};

fsa.prototype.parse =  function(){
  let _state_info = [];
  let _trans = this.transition;
  let state = this.start_state;
  let _fin = this.fin_state;
  let _speech = this.speech_list;

  let input_str = arguments[0].toString().split(" ");


  for(let i = 0 ; i < input_str.length ;i++){
    let info = new String("");
    info = info + ("Current state : " + state);
    state = _trans[state][_speech[input_str[i]]];
    info = info + (" Next state : " + state);

    _state_info.push(info);
  }

  if(Array.isArray(_fin)){
    for(let i of _fin)
      if(state == i){
        _state_info.push("Done!");
        return _state_info;
      }
  }
  else{
    if(state == _fin){
      _state_info.push("Done!");
      return _state_info;
    }
  }
  _state_info.push("Failed!");
  return _state_info;
};

File: NLP_HW1/flask/fsa.py
class fsa:
    def __init__(self , translate , start_state , fin_state , speech_list):
        self.translate = translate
        self.start_state = start_state
        self.fin_state  = fin_state
        self.speech_list = speech_list

    def parse(self , input_str):
        state_info = []
        state = self.start_state
        for i in input_str:
            info = 'Current state : %s ' % (state)
            part_of_speech = self.speech_list[i]
            state = self.translate[state][part_of_speech]
            info += 'and next state : %s' % (state)
            state_info.append(info)

        if isinstance(self.fin_state , tuple):
            for i in self.fin_state:
                if state == i:
                    state_info.append('Done!')
                    return state_info
        else:
            if state == self.fin_state:
                state_info.append('Done!')
                return state_info
        state_info.append('Failed!')
        return state_info
